app/Http/Permissions/PermissionMiddleware.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace App\Http\Permissions;
+
+use App\Exceptions\ApiException;
+use App\Exceptions\ExceptionFactory;
+use App\Models\User;
+use Closure;
+use Illuminate\Http\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class PermissionMiddleware
+{
+    public const PERMISSIONS_KEY = 'permissions-key';
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param Request $request
+     * @param Closure(Request): (Response) $next
+     * @param string ...$permissions
+     * @return Response
+     * @throws ApiException
+     */
+    public function handle(Request $request, Closure $next, string  ...$permissions): Response
+    {
+        $permissionObject = $this->requestPermissions($request);
+        $passed = false;
+        foreach ($permissions as $permissionCode) {
+            //
+            $permission = Permission::fromName($permissionCode);
+            if ($permissionObject->getByPermission($permission)) {
+                $passed = true;
+                break;
+            }
+        }
+        if (!$passed) {
+            if ($permissionObject->getByPermission(Permission::unauthorised)) {
+                throw ExceptionFactory::unauthorised();
+            }
+            throw ExceptionFactory::forbidden();
+        }
+        return $next($request);
+    }
+
+    private function requestPermissions(Request $request): PermissionObject
+    {
+        $attributes = $request->attributes;
+        $permissionObject = $attributes->get(self::PERMISSIONS_KEY);
+        if ($permissionObject instanceof PermissionObject) {
+            return $permissionObject;
+        }
+        $unauthorised = false;
+        $unverified = false;
+        $verified = false;
+        $globalAdmin = false;
+
+        /** @var User|null $user */
+        $user = $request->user();
+        if ($user) {
+            $verified = ($user->email_verified_at !== null);
+            $unverified = !$verified;
+            $globalAdmin = ($user->global_admin_grant_id !== null);
+        } else {
+            $unauthorised = true;
+        }
+        $permissionObject = new PermissionObject(
+            unauthorised: $unauthorised,
+            unverified: $unverified,
+            verified: $verified,
```

Comment:
tu trochę nie rozumiem - nie wystarczy sam property `verified`? ;d

Reply:
chodzi o to, że może być endpoint czy strona do weryfikacji email, która wymaga żeby być unverified, a sprawdzanie uprawnień chcę zawsze robić, że sprawdzam czy jest true (w tym przypadku to by było `authorised && !verifiued`, to wole `unverified`)

i też jak chce zdobić stronę do której ma dostęp zalogowany użytkownik nawet jak nie ma zweryfikowanego adresu email, to musze to jasno zadeklarować przez `$this->permissionOneOf(Permission::unverified, Permission::verified)`, trudniej dać przypadkiem za duży dostęp